Show latest article of the DBlog on the frontpage

wilzbach opened the pull request:
I just had this crazy idea and funnily it works quite well - even with Ddoc :)

This definition is only used when `DIFFABLE` is not set, so it won't be possible to preview this on the DAutoTest, but here's how it will look in action:

![image](https://user-images.githubusercontent.com/4370550/27181491-17809650-51d8-11e7-907e-da80aa11c065.png).

For a local build without internet, `curl` will fail silently and it will just create an empty file.
On subsequent runs, this target won't be executed. (No `.PHONY`).

This approach is also nicer than the previous JavaScript injection, because search engines see the new content & they like sites that are updated often.

It's fairly unlikely that the DBlog theme will change soon and if - as said before - it will just silently fail ...

CC @mdparker @JackStouffer @CyberShadow

ZombineDev commented on index.dd:
```diff
@@ -94,10 +94,11 @@ $(DIVC boxes,
-            $(P Stay updated with
-                $(LINK2 http://arsdnet.net/this-week-in-d,
-                    <cite>This Week in D</cite>) -
-                $(SPANC twid)
```
Let's keep This Week in D. It's pretty cool, despite not having the "official" stamp on it. Just move it below the official blog. May be something like:

> Stay updated with the [official D Blog](https://dlang.org/blog/) and [This Week in D](http://arsdnet.net/this-week-in-d):
> * [D Blog - Compile-Time Sort in D](http://dlang.org/blog/2017/06/05/compile-time-sort-in-d/) by Michael Parker from June 5, 2017.
> * [This Week in D - May 28](http://arsdnet.net/this-week-in-d/2017-may-28.html): Blog article on optimization and Andrei Israel trip, Beta D 2.074, tip on mixin w/ anon func.

wilzbach commented:
> Let's keep This Week in D. It's pretty cool, despite not having the "official" stamp on it. Just move it below the official blog. May be something like:

I considered this for a moment, but then dropped it because of space problems:

![image](https://user-images.githubusercontent.com/4370550/27187061-7428352c-51ea-11e7-8bda-899104349e4b.png)

Any ideas?

adamdruppe commented:
I didn't get a chance to update twid again since I was out of town again.... I wouldn't mind if you simplified or removed it to save space.

ZombineDev commented:
If space is an issue we can try with this:

> Stay updated with:
> * The official D Blog: [Compile-Time Sort in D](http://dlang.org/blog/2017/06/05/compile-time-sort-in-d/) from June 5 2017, and
> * This Week in D:  [May 28](http://arsdnet.net/this-week-in-d/2017-may-28.html).

Or with an even shorter version:

> Stay updated with the [official D Blog - Compile-Time Sort in D](http://dlang.org/blog/2017/06/05/compile-time-sort-in-d/) from June 5 2017 and [This Week in D - May 28](http://arsdnet.net/this-week-in-d/2017-may-28.html).

wilzbach commented:
CC @aG0aep6G - as you rescued the Dlang.org template - what's your opinion on this?

CC @popen2 - at DConf you mentioned that you were planning on redesigning dlang.org - maybe you have already an idea in mind?

aG0aep6G commented:
> CC @aG0aep6G - as you rescued the Dlang.org template - what's your opinion on this?

I really dislike how much the home page currently depends on JavaScript, so this is interesting. It feels weird to generate the "latest" stuff when building the site, but if you do that regularly enough, sure.

Does it work, though? The doc tester shows "from : by . ". Also, https://blog.dlang.org doesn't work. Needs "http://".

Regarding space problems, you could cut the DConf text down. One paragraph instead of two; shorter "thank you" or none at all; "through this YouTube playlist" -> "on YouTube" - stuff like that.

Regarding TWID, it doesn't seem very alive. So I'd go with the blog if only one can get in.

CyberShadow commented:
I think the DConf bit needs to be moved out of the news section and somewhere else, a banner or a sidebar widget. Then there should be enough room for the blog and TWiD in the News section (no need to listify them either, just two paragraphs).

wilzbach commented:
> Does it work, though? The doc tester shows "from : by . ".

Yes as mentioned before, I explicitly disable this (as dynamic content) on the DAutoTest to avoid random changes (`DIFFABLE=1` is set on DAutotTest).
However on the deployment, `DIFFABLE` isn't set. For example, the [pending changelog](https://dlang.org/changelog/2.075.0_pre.html) is built similarly.
If the `curl` trick looks too weird, I can always plugin a simple DScript instead...

> Also, https://blog.dlang.org doesn't work. Needs "http://".

Oh thanks!

> I think the DConf bit needs to be moved out of the news section and somewhere else, a banner or a sidebar widget. Then there should be enough room for the blog and TWiD in the News section (no need to listify them either, just two paragraphs).

Okay,
- I also parse TWID at "Compile-Time" now, so there's one ugly layout reflow less on the main page.
- I removed the "Thank you DConf 2017" notice 

![image](https://user-images.githubusercontent.com/4370550/27207778-904d050a-5241-11e7-96ec-9ad5dead4128.png)

Now adding the DConf link to "Learn" yields (and slightly adjusting the News intro):

![image](https://user-images.githubusercontent.com/4370550/27208006-503bfef6-5243-11e7-9ec4-15cc04c6fc3c.png)

@CyberShadow the only place I could find to put the playlist link was the "Learn" section. Better ideas?

CyberShadow commented:
> Better ideas?

I was thinking maybe take http://dconf.org/2017/images/dconf_logo_2017.svg, crop it, and put it in the sidebar above or under the Tweeter widget, what do you think? Plenty of space there.

wilzbach commented:
> I was thinking maybe take http://dconf.org/2017/images/dconf_logo_2017.svg, crop it, and put it (with some text) in the sidebar above or under the Tweeter widget, what do you think? Plenty of space there.

Nice idea - like this? Maybe a bit more spacing?

![image](https://user-images.githubusercontent.com/4370550/27265180-04fd792e-5491-11e7-99c0-090cc59053ca.png)

CyberShadow commented:
Yep! Could use a bit of text under the logo (essentially what we have in the news section right now).

wilzbach commented:
> Yep! Could use a bit of text under the logo (essentially what we have in the news section right now).

default

![image](https://user-images.githubusercontent.com/4370550/27265246-3e69e480-5492-11e7-9159-52a966fc7c8a.png)

12px

![image](https://user-images.githubusercontent.com/4370550/27265262-5def84fe-5492-11e7-9c77-25ead8c7f413.png)

CyberShadow commented on sitemap.dd:
```diff
@@ -0,0 +1,203 @@
+Ddoc
```
Did you add this accidentally?

wilzbach replied:
Yes, I really need some sleep :P
In any case, we should add this to the gitignore or even better: simply put it into `.generated`
https://github.com/dlang/dlang.org/pull/1725

CyberShadow commented on posix.mak:
```diff
@@ -378,6 +383,22 @@ dlangspec.verbatim.txt : $(DMD) verbatim.ddoc dlangspec-consolidated.d
+################################################################################
+
+$G/dblog_latest.ddoc:
+	@echo "Receiving the latest DBlog article. Disable with DIIFABLE=1"
```
DIFFABLE, not DIIFABLE

CyberShadow commented on posix.mak:
```diff
@@ -378,6 +383,22 @@ dlangspec.verbatim.txt : $(DMD) verbatim.ddoc dlangspec-consolidated.d
+$G/twid_latest.ddoc:
+	@echo "Receiving the latest TWID article. Disable with DIIFABLE=1"
+	curl -s --retry 3 --retry-delay 5 http://arsdnet.net/this-week-in-d/twid-latest.js | grep setTwid | \
+		sed -E 's/.*setTwid\("([^"]+)", "([^"]+)", "([^"]+)"\).*/TWID_LATEST_LINK=\1|TWID_LATEST_DATE=\2|TWID_LATEST_TITLE=\3/' | \
```
Is this going to handle special characters nicely?

CyberShadow commented on images/dconf_logo_2017.svg:
```diff
@@ -0,0 +1,213 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Hotlinking it would've been fine, but this is fine too

wilzbach replied:
I prefer copying - just in case the layout of dconf.org changes ...

wilzbach commented on index.dd:
```diff
@@ -128,7 +142,10 @@ $(DIVC boxes,
+                <cite>Programming in D</cite>) or
+                $(LINK2 http://youtube.com/playlist?list=PL3jwVPmk_PRxo23yyoc0Ip_cP3-rCm7eB, watch)
+                talks from
+                $(LINK2 http://dconf.org/2017/index.html, DConf17)
```
Should I still keep this if we add the DConf banner?

CyberShadow replied:
Yeah, not sure how good the talks are effective for actually learning D.

wilzbach commented:
> Is this going to handle special characters nicely?

- I don't see why not? [Example with Regex101](https://regex101.com/r/cGz0vE/1/)
- [RFC 3966](https://tools.ietf.org/html/rfc3986#section-2) defines `|` as an invalid symbol and would need to be escaped

CyberShadow commented:
If I understand that regex correctly, if a TWiD issue has a " in it, it's not going to show up at all.

CyberShadow commented:
I don't understand how RFC 3966 plays in this. The TWiD data is a JavaScript file, it's not URL-encoded.

wilzbach commented:
> If I understand that regex correctly, if a TWiD issue has a " in it, it's not going to show up at all.
> don't understand how RFC 3966 plays in this. The TWiD data is a JavaScript file, it's not URL-encoded.

Yes & Yes, sorry.
Updated the RegEx: https://regex101.com/r/cGz0vE/2 (at least for TWID I hope that this is only a temporary solution anyways as for @adamdruppe it's probably the same effort to write his latest article information in this extracted Ddoc format than JS).

wilzbach commented:
Alrighty, I have fixed all open "issues" and this should be good to go. Anything else blocking this?
(I also squashed the DConf banner changes into one commit).

CyberShadow commented:
@adamdruppe Any news on twid_latest.dd?

CyberShadow commented on posix.mak:
```diff
@@ -378,6 +383,22 @@ dlangspec.verbatim.txt : $(DMD) verbatim.ddoc dlangspec-consolidated.d
+$G/twid_latest.ddoc:
+	@echo "Receiving the latest TWID article. Disable with DIFFABLE=1"
+	curl -s --retry 3 --retry-delay 5 http://arsdnet.net/this-week-in-d/twid-latest.js | grep setTwid | \
+		sed -E 's/.*setTwid\("([^"]+)", "([^"]+)", "(.+)"\);$$/TWID_LATEST_LINK=\1|TWID_LATEST_DATE=\2|TWID_LATEST_TITLE=\3/' | \
```
Quotes (and other JS escapes) in TWID titles are still going to show up wrong.

adamdruppe commented:
http://arsdnet.net/this-week-in-d/twid_latest.dd

wilzbach commented:
Thanks a lot @adamdruppe! Integrated & rebased. Anything else missing? :)

wilzbach commented:
For reference, this is how it currently looks:

![image](https://user-images.githubusercontent.com/4370550/27603801-99623270-5b77-11e7-9074-fb7577dd2206.png)

![image](https://user-images.githubusercontent.com/4370550/27603819-a7685840-5b77-11e7-87b7-d15c106ce245.png)

CyberShadow commented:
Looks great!

wilzbach commented:
And fetching the latest resources on the deploy for dlang.org worked - it's live now :)

Thanks a lot @adamdruppe @mdparker @ZombineDev @aG0aep6G @CyberShadow !

@adamdruppe dlang.org gets built and deployed everytime a PR gets merged to dlang/{dmd,druntime,phobos,dlang.org} - according to your stats that's about 30 times a week or ~ three times a day. So, in the worst case, there might be a lag of a few hours until new articles are propagated on dlang.org.

CyberShadow commented:
If you want, I can add a trigger that forces a rebuild

adamdruppe commented:
On Tue, Jun 27, 2017 at 11:58:35AM -0700, Sebastian Wilzbach wrote:
> @adamdruppe dlang.org gets built and deployed everytime a PR gets merged to dlang/{dmd,druntime,phobos,dlang.org}

There's a lag with me updating it too, some days I don't get around
to it till late monday, so not a big deal.
